torch/distributed/device_mesh.py

```diff
@@ -73,32 +73,32 @@ def get_current_mesh(self) -> "DeviceMesh":
             return self.mesh_stack[-1]
 
         def create_child_mesh(
```

Comment:
rename to `create_sub_mesh`

Also: we should carefully check inside this function to avoid creating a submesh from non-root mesh

Reply:
With the change of recording child_to_root mesh relationship, we would correctly assign the root mesh even if a submesh is created from a non-root mesh as shown in the second example in the summary. 

Do we still want to make it more strict and disallow the usage of creating submesh from non-root mesh completely?

Reply:
I think we should always start with more restrictions, and only loosen the restriction if there's a reason or user requested it

Reply:
> I think we should always start with more restrictions, and only loosen the restriction if there's a reason or user requested it

One more thing is that this usage of slicing on a submesh is currently allowed, although there might not be lots of usages yet, but this is bc breaking if we were to restrict it though.

Reply:
I don't think it's going to be a problem for two reasons: 1. even though it's allowed, it's not sth user would typically use (i.e. I couldn't think of a reason why user want to use it that way) 2. DeviceMesh currently mainly having the tutorial/recipe, where we didn't document this behavior explicitly yet anywhere.

If you are concerned about this, we can also file a BC breaking but I feel it might not be necessary

Reply:
> I don't think it's going to be a problem for two reasons: 1. even though it's allowed, it's not sth user would typically use (i.e. I couldn't think of a reason why user want to use it that way) 2. DeviceMesh currently mainly having the tutorial/recipe, where we didn't document this behavior explicitly yet anywhere.
> 
> If you are concerned about this, we can also file a BC breaking but I feel it might not be necessary

If we are gonna restrict this, we can raise an error if a slice is called upon a submesh, but I think we should expose an API like `is_submesh` in `DeviceMesh`. As the child to root relationship is only recorded in the _MeshEnv, users actually can't tell whether a `DeviceMesh` is created through slicing or not.

Reply:
hmmm I don't understand, why user need to tell this? Could you explain more? IMO we should just have clear error messages to user that they CANNOT create a submesh from a submesh, because we know exactly the root to sub mesh relationship when user creating a submesh

Reply:
My thought on this `is_submesh` API is that we still don't want to encode any submesh-rootmesh relationship into DeviceMesh itself (i.e. the same reason we removed parent_mesh from DeviceMesh as attribute), as DeviceMesh public API should be mainly about the DeviceMesh itself, and the `_MeshEnv` responsible for maintaining device mesh relationships.

We can probably think about how/whether we should make MeshEnv be public, but that would require proper designs about the API interface exposing to the user, etc.

Reply:
> hmmm I don't understand, why user need to tell this? Could you explain more? IMO we should just have clear error messages to user that they CANNOT create a submesh from a submesh, because we know exactly the root to sub mesh relationship when user creating a submesh

Ye. We would have this information because we recorded child to root_mesh relationship in _MeshEnv so it's not an issue for us. 

For users writing their own distributed program and the initialization could happen in multiple places. They might not necessarily remember which variable is a submesh and which variable is not a submesh. If we throw an error message restricting the usage, it might be helpful for users to be able to find out which mesh is a submesh and which mesh is not. Just some thoughts coming from UX perspective.

Reply:
I think from UX prospective is also not good to expose such an API, as this would again bundle device mesh itself with the device mesh relationships. This would give additional complexity to user that they need to understand what happened under the hood, but it does not provide full functionality that advanced users need, for example, advanced users might want to not only know whether this is a submesh, but also whether is a submesh of a certain mesh, what's the root mesh of this submesh, all of these informations require a separate API, which if we expose all of them then they are just MeshEnv

Reply:
Makes sense. 

Updated to add an additional check for creating submesh here: 
https://github.com/pytorch/pytorch/blob/7d33426f06d8ff52c5954d21099acb206f13e31e/torch/distributed/device_mesh.py#L78-L79